setup.py

```diff
@@ -1191,64 +1197,70 @@ class CythonCommand(build_ext_c):
         def build_extension(self, ext):
             pass
 
-    class zbuild_ext(build_ext_c):
+    class zbuild_ext(build_ext_cython):
 
         def build_extensions(self):
             if self.compiler.compiler_type == 'mingw32':
                 customize_mingw(self.compiler)
-            return build_ext_c.build_extensions(self)
+            return super().build_extensions()
 
         def build_extension(self, ext):
             with fix_cxx(self.compiler, ext):
-                build_ext.build_extension(self, ext)
+                super().build_extension(ext)
             ext_path = self.get_ext_fullpath(ext.name)
             patch_lib_paths(ext_path, self.compiler.library_dirs)
 
-        def run(self):
-            self.distribution.run_command('configure')
-
-            return build_ext_c.run(self)
+        def finalize_options(self):
+            self.distribution.run_command("configure")
+            return super().finalize_options()
 
-    cmdclass['cython'] = CythonCommand
-    cmdclass['build_ext'] =  zbuild_ext
+    cmdclass["cython"] = CythonCommand
+    cmdclass["build_ext"] = zbuild_ext
 
 extensions = []
 ext_include_dirs = [pjoin('zmq', sub) for sub in ('utils',)]
 ext_kwargs = {
     'include_dirs': ext_include_dirs,
 }
-if cython:
-    # set binding so that compiled methods can be inspected
-    ext_kwargs['cython_directives'] = {'binding': True}
-    if cython_language_level:
-        ext_kwargs['cython_directives']['language_level'] = cython_language_level
 
 for submod, packages in submodules.items():
     for pkg in sorted(packages):
-        sources = [pjoin('zmq', submod.replace('.', os.path.sep), pkg+suffix)]
-        ext = Extension(
-            'zmq.%s.%s'%(submod, pkg),
-            sources = sources,
-            **ext_kwargs
-        )
+        sources = [pjoin("zmq", submod.replace(".", os.path.sep), pkg + suffix)]
+        ext = Extension("zmq.%s.%s" % (submod, pkg), sources=sources, **ext_kwargs)
         extensions.append(ext)
 
+if cython:
+    # set binding so that compiled methods can be inspected
+    # set language-level to 3str, requires Cython 0.29
+    cython_directives = {"binding": True, "language_level": "3str"}
+    extensions = cythonize(extensions, compiler_directives=cython_directives)
+
 if pypy:
     # add dummy extension, to ensure build_ext runs
-    dummy_ext = Extension('dummy', sources=[])
+    dummy_ext = Extension("dummy", sources=[])
     extensions = [dummy_ext]
 
     bld_ext = cmdclass['build_ext']
+
     class pypy_build_ext(bld_ext):
         """hack to build pypy extension only after building bundled libzmq
 
         otherwise it will fail when libzmq is bundled.
         """
-        def build_extensions(self):
-            self.extensions.remove(dummy_ext)
-            bld_ext.build_extensions(self)
+
+        def finalize_options(self):
+            super().finalize_options()
```

Comment:
@minrk this gives me an error when building on Debian testing:

```
[  400s] running build_ext
[  400s] Traceback (most recent call last):
[  400s]   File "setup.py", line 1374, in <module>
[  400s]     setup(**setup_args)
[  400s]   File "/usr/lib/pypy/dist-packages/setuptools/__init__.py", line 162, in setup
[  400s]     return distutils.core.setup(**attrs)
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/core.py", line 151, in setup
[  400s]     dist.run_commands()
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/dist.py", line 953, in run_commands
[  400s]     self.run_command(cmd)
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/dist.py", line 972, in run_command
[  400s]     cmd_obj.run()
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/command/build.py", line 127, in run
[  400s]     self.run_command(cmd_name)
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/cmd.py", line 334, in run_command
[  400s]     self.distribution.run_command(command)
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/dist.py", line 971, in run_command
[  400s]     cmd_obj.ensure_finalized()
[  400s]   File "/usr/lib/pypy/lib-python/2.7/distutils/cmd.py", line 109, in ensure_finalized
[  400s]     self.finalize_options()
[  400s]   File "setup.py", line 1252, in finalize_options
[  400s]     super().finalize_options()
[  400s] TypeError: __init__() takes at least 2 arguments (1 given)
[  400s] E: pybuild pybuild:352: build: plugin distutils failed with: exit code=1: /usr/bin/pypy setup.py build 
```

Any idea?

Reply:
yes, that's Python 2.7 which is no longer supported.

Reply:
Uh weird, I removed all the python2 stuff. Maybe it's the shbang in the setup.py?

Reply:
Ah it was pypy - didn't notice there's a separate pypy3 package to use